ann_benchmarks/algorithms/milvus.py

```diff
@@ -1,45 +1,71 @@
 from __future__ import absolute_import
-import milvus
+from sqlite3 import paramstyle
+from pymilvus import (
+    connections,
+    utility,
+    FieldSchema,
+    CollectionSchema,
+    DataType,
+    IndexType,
+    Collection,
+)
 import numpy
 import sklearn.preprocessing
 from ann_benchmarks.algorithms.base import BaseANN
 
 
 class Milvus(BaseANN):
-    def __init__(self, metric, index_type, nlist):
-        self._nlist = nlist
+    def __init__(self, metric, conn_params, index_type, method_params):
+        self._host = conn_params['host']
+        self._port = conn_params['port'] # 19530
+        # connections.connect(host=conn_params['host'], port=conn_params['port'])
+        # fields = [
+        #     FieldSchema(name="pk", dtype=DataType.INT64, is_primary=True, auto_id=False),
+        #     FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=100)
+        # ]
+        # schema = CollectionSchema(fields)
+        # self._milvus = Collection('milvus', schema)
+        self._index_type = index_type
+        self._method_params = method_params
         self._nprobe = None
         self._metric = metric
-        self._milvus = milvus.Milvus()
-        self._milvus.connect(host='localhost', port='19530')
-        self._table_name = 'test01'
-        self._index_type = index_type
 
     def fit(self, X):
         if self._metric == 'angular':
-            X = sklearn.preprocessing.normalize(X, axis=1, norm='l2')
+            X = sklearn.preprocessing.normalize(X, axis=1)
+        
+        # TODO: if we can set the dim later, mabe return this to the init func
+        connections.connect(host=self._host, port=self._port)
+        fields = [
+            FieldSchema(name="pk", dtype=DataType.INT64, is_primary=True, auto_id=False),
+            FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=len(X[0]))
+        ]
+        schema = CollectionSchema(fields)
+        self._milvus = Collection('milvus', schema)
 
-        self._milvus.create_table({'table_name': self._table_name, 'dimension': X.shape[1]})
-        vector_ids = [id for id in range(len(X))]
-        self._milvus.insert(table_name=self._table_name, records=X.tolist(), ids=vector_ids)
-        index_type = getattr(milvus.IndexType, self._index_type)  # a bit hacky but works
-        self._milvus.create_index(self._table_name, {'index_type': index_type, 'nlist': self._nlist})
+        self._milvus.insert([[id for id in range(len(X))], X.tolist()])
+        self._milvus.create_index('vector', {'index_type': self._index_type, 'metric_type':'L2', 'params':self._method_params})
```

Comment:
are `M` and `ef` are method params here for HNSW?
more general: does `self_.method_params` fit every use case of milvus?

Reply:
yes. milvus expects a dictionary that contains mandatory parameters, different for every algorithm. this way I can set the params dictionary in algo.yaml for every algorithm and not have to fit the functions to every possible parameters